snapcraft/internal/elf.py

```diff
@@ -137,8 +138,12 @@ def _crawl_for_path(*, soname: str, root_path: str,
                 if file_name == soname:
                     file_path = os.path.join(root, file_name)
                     if ElfFile.is_elf(file_path):
-                        soname_cache[soname] = file_path
-                        return file_path
+                        # We found a match by name, anyway. Let's verify that
+                        # the architecture is the one we want.
+                        elf_file = ElfFile(path=file_path)
+                        if elf_file.arch == arch:
+                            soname_cache[soname] = file_path
```

Comment:
One concern here is that if the same SonameCache instance gets reused to crawl dependencies of two ELF files of different architectures, the second will match entries added by the first.

One way to handle this would be to make the keys in the SonameCache be `(arch, soname)`.  This way it doesn't matter if cache entries are added for other architectures because you'll never look up those keys.

Reply:
Oh great catch, that would have been one nasty bug.